### PR TITLE
MT3-250 #done #comment Add Fieldset component

### PR DESCRIPTION
**Pull request by CJWaszczuk**
# What?

Added fieldset and fieldset legend components

# Why?

One or more elements within a form may need to be grouped.

# Notes
Relies on a version increase of `lbh-frontend` so that the extends are removed and the CSS doesn't break. 
 

# References

`lbh-frontend` library: [Github](https://github.com/LBHackney-IT/LBH-frontend/tree/master/src/lbh/components/lbh-fieldset) | [Style guide](https://lbh-frontend.herokuapp.com/components/lbh-fieldset)
`govuk-frontend` library : [Github](https://github.com/alphagov/govuk-frontend/tree/master/src/govuk/components/fieldset) | [Style guide](https://design-system.service.gov.uk/components/fieldset/)

### CODE REVIEW

**Comment by erbridge** [comment]
If you're going to create draft PRs, please could you note why they're drafts in the description.

**Comment by CJWaszczuk** [comment]
@erbridge I think there's a question here about whether we want `FieldsetLegend` to actually be an exported thing given that it's use case is `Fieldset` only.

File: src/components/Fieldset.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  className?: string;
+  isPageHeading?: boolean;
+  children?: React.ReactNode;
+}
+
+/**
+ * The type of a valid fieldset legend component. It allows any React functional
+ * components that accept {@link FieldsetLegendProps} as proptypes.
+ */
+export type FieldsetLegendType = React.FunctionComponent<FieldsetLegendProps>;
```
Comment (erbridge) [inline src/components/Fieldset.tsx]: Does this get used anywhere outside of this file? If not, I don't think we need the alias.

File: src/components/Fieldset.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+/**
+ * The proptypes for {@link Fieldset}.
+ */
+export interface FieldsetProps {
+  id?: string;
+  className?: string;
+  role?: string;
+  fieldsetLegend?: JSX.Element;
```
Comment (erbridge) [inline src/components/Fieldset.tsx]:
```suggestion
  legend?: JSX.Element;
```

No need to repeat ourselves.

File: src/components/Fieldset.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * The type of a valid fieldset legend component. It allows any React functional
+ * components that accept {@link FieldsetLegendProps} as proptypes.
+ */
+export type FieldsetLegendType = React.FunctionComponent<FieldsetLegendProps>;
+
+/**
+ * The legend for {@link Fieldset}.
+ */
+export const FieldsetLegend: FieldsetLegendType = (
```
Comment (erbridge) [inline src/components/Fieldset.tsx]: If we didn't export this, how would a user create one?

File: src/components/Fieldset.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  id?: string;
+  className?: string;
+  role?: string;
+  fieldsetLegend?: JSX.Element;
+  describedBy?: string;
+  children: React.ReactNode;
+}
+
+/**
+ * A fieldset used to group elements within a form.
```
Comment (erbridge) [inline src/components/Fieldset.tsx]:
```suggestion
 * A component used to group elements within a form.
```

The name of the component is `Fieldset`, so repetition is redundant.

File: src/components/Fieldset.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+/**
+ * The type of a valid fieldset legend component. It allows any React functional
+ * components that accept {@link FieldsetLegendProps} as proptypes.
+ */
+export type FieldsetLegendType = React.FunctionComponent<FieldsetLegendProps>;
+
+/**
+ * The legend for {@link Fieldset}.
```
Comment (erbridge) [inline src/components/Fieldset.tsx]:
```suggestion
 * A component to set a legend for a {@link Fieldset} by passing a {@link FieldsetProps.legend}.
```

Also include a usage example either here or in `Fieldset`'s docstring, so demonstrate how to provide a legend in addition to some inputs.

**Review by erbridge — COMMENTED** [review]
I don't understand your question about not needing to export `FieldsetLegend`. See my comment at that point.

I agree that it makes sense to add `lbh-fieldset__legend` to the legend even if it's not currently used.

File: src/components/Fieldset.tsx (hunk 3 above)
Comment (CJWaszczuk) [reply src/components/Fieldset.tsx]: Resolved in conversation, my comment didn't make sense.

File: src/components/Fieldset.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ *     <FieldsetLegend id="legendId" className="legendClass" isPageHeading>
+ *       children
+ *     </FieldsetLegend>
+ *   }
+ * >
+ *    children
+ * </Fieldset>
+ * ```
+ */
+export const FieldsetLegend: React.FunctionComponent<FieldsetLegendProps> = (
```
Comment (erbridge) [inline src/components/Fieldset.tsx]: Let's do the same thing as with buttons and pull these into separate files in their own folder. Feels a bit messy to have both components in here.

File: src/components/Fieldset.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  props: FieldsetLegendProps
+): JSX.Element => {
+  const { id, className, isPageHeading, children } = props;
+  return (
+    <legend id={id} className={classNames("govuk-fieldset__legend", className)}>
```
Comment (erbridge) [inline src/components/Fieldset.tsx]:
```suggestion
    <legend id={id} className={classNames("govuk-fieldset__legend", "lbh-fieldset__legend", className)}>
```

Let's follow BEM even if there aren't any rules for the class (yet).

File: src/components/Fieldset.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      {isPageHeading ? (
+        <h1 className="govuk-fieldset__heading">{children}</h1>
```
Comment (erbridge) [inline src/components/Fieldset.tsx]:
```suggestion
        <h1 className="govuk-fieldset__heading lbh-fieldset__heading">{children}</h1>
```

As above.

**Comment by CJWaszczuk** [comment]
There's no explicit stylesheet for fieldset legends, so we'll end up having to duplicate the the fieldset css across both components.

**Comment by erbridge** [comment]
It's not duplication. Node knows how to deduplicate imports.

File: src/components/Fieldset/FieldsetLegend.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+export interface FieldsetLegendProps {
+  id?: string;
+  className?: string;
+  isPageHeading?: boolean;
+  children: React.ReactNode;
+}
+
+/**
+ * A component to set a legend for a {@link Fieldset} by passing
+ * a {@link FieldsetProps.legend} prop.
```
Comment (erbridge) [inline src/components/Fieldset/FieldsetLegend.tsx]:
```suggestion
 * in via the {@link FieldsetProps.legend} prop.
```

I know I wrote this, but it reads weirdly.

File: src/components/Fieldset/FieldsetLegend.tsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+import "lbh-frontend/lbh/components/lbh-fieldset/_fieldset.scss";
+
+/**
+ * The proptypes for {@link FieldsetLegend}.
+ */
+export interface FieldsetLegendProps {
+  id?: string;
+  className?: string;
+  isPageHeading?: boolean;
```
Comment (erbridge) [inline src/components/Fieldset/FieldsetLegend.tsx]: Comment this prop, please. What effect does it have?

File: src/components/Fieldset/FieldsetLegend.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+ *   className="fieldsetClass"
+ *   describedBy="fieldsetDescribedBy"
+ *   role="fieldsetRole"
+ *   legend={
+ *     <FieldsetLegend id="legendId" className="legendClass" isPageHeading>
+ *       children
+ *     </FieldsetLegend>
+ *   }
+ * >
+ *    children
```
Comment (erbridge) [inline src/components/Fieldset/FieldsetLegend.tsx]:
```suggestion
 *   children
```

File: src/components/Fieldset/Fieldset.tsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import "lbh-frontend/lbh/components/lbh-fieldset/_fieldset.scss";
+
+/**
+ * The proptypes for {@link Fieldset}.
+ */
+export interface FieldsetProps {
+  id?: string;
+  className?: string;
+  role?: string;
+  legend?: JSX.Element;
```
Comment (erbridge) [inline src/components/Fieldset/Fieldset.tsx]: Document what this is supposed to be, please.

**Review by erbridge — APPROVED** [review]

**Pull request merged**